sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -143,145 +115,231 @@ private[sql] object OrcFilters extends OrcFiltersBase {
+            ()
+        }
       case GreaterThanOrEqual(attribute, value) if isSearchableType(dataTypeMap(attribute)) =>
-        val quotedName = quoteAttributeNameIfNeeded(attribute)
-        val castedValue = castLiteralValue(value, dataTypeMap(attribute))
-        Some(builder.startNot().lessThan(quotedName, getType(attribute), castedValue).end())
-
+        actionType match {
+          case _: TrimUnconvertibleFilters => Some(expression)
+          case BuildSearchArgument(builder) =>
+            val quotedName = quoteAttributeNameIfNeeded(attribute)
+            val castedValue = castLiteralValue(value, dataTypeMap(attribute))
+            builder.startNot().lessThan(quotedName, getType(attribute), castedValue).end()
+            ()
+        }
       case IsNull(attribute) if isSearchableType(dataTypeMap(attribute)) =>
-        val quotedName = quoteAttributeNameIfNeeded(attribute)
-        Some(builder.startAnd().isNull(quotedName, getType(attribute)).end())
-
+        actionType match {
+          case _: TrimUnconvertibleFilters => Some(expression)
+          case BuildSearchArgument(builder) =>
+            val quotedName = quoteAttributeNameIfNeeded(attribute)
+            builder.startAnd().isNull(quotedName, getType(attribute)).end()
+            ()
+        }
       case IsNotNull(attribute) if isSearchableType(dataTypeMap(attribute)) =>
-        val quotedName = quoteAttributeNameIfNeeded(attribute)
-        Some(builder.startNot().isNull(quotedName, getType(attribute)).end())
-
+        actionType match {
+          case _: TrimUnconvertibleFilters => Some(expression)
+          case BuildSearchArgument(builder) =>
+            val quotedName = quoteAttributeNameIfNeeded(attribute)
+            builder.startNot().isNull(quotedName, getType(attribute)).end()
+            ()
+        }
       case In(attribute, values) if isSearchableType(dataTypeMap(attribute)) =>
-        val quotedName = quoteAttributeNameIfNeeded(attribute)
-        val castedValues = values.map(v => castLiteralValue(v, dataTypeMap(attribute)))
-        Some(builder.startAnd().in(quotedName, getType(attribute),
-          castedValues.map(_.asInstanceOf[AnyRef]): _*).end())
+        actionType match {
+          case _: TrimUnconvertibleFilters => Some(expression)
+          case BuildSearchArgument(builder) =>
+            val quotedName = quoteAttributeNameIfNeeded(attribute)
+            val castedValues = values.map(v => castLiteralValue(v, dataTypeMap(attribute)))
+            builder.startAnd().in(quotedName, getType(attribute),
+              castedValues.map(_.asInstanceOf[AnyRef]): _*).end()
+            ()
+        }
 
-      case _ => None
+      case _ =>
+        actionType match {
+          case _: TrimUnconvertibleFilters => None
+          case BuildSearchArgument(builder) =>
+            throw new IllegalArgumentException(s"Can't build unsupported filter ${expression}")
+        }
     }
   }
+
+  private def performFilter(
```

Comment:
I think it's good to keep `performFilter`. BTW we can remove the  `canPartialPushDownConjuncts` because the only caller always set this parameter to true.

Reply:
or, we can remove both `performFilter` and `updateBuilder`.

Reply:
I removed the `performFilter` method and effectively replaced it with the `trimUnconvertibleFilters` one. I prefer to keep `updateBuilder` because I think it gives the right level of abstraction for the overall `buildSearchArgument` method: trim, and then build. Without these helped methods, it starts looking like this:

```
    performAction(TrimUnconvertibleFilters(canPartialPushDownConjuncts = true), expression)
        .map { filter =>
          performAction(BuildSearchArgument(builder), filter)
          builder
        }
```

which is a bit too noisy and convoluted to grasp the high-level structure at a glance.